address/src/derive.rs

```diff
@@ -0,0 +1,142 @@
+use crate::{Address, MAX_SEEDS, PDA_MARKER};
+#[cfg(target_os = "solana")]
+use {core::mem::MaybeUninit, solana_define_syscall::definitions::sol_sha256};
+
+/// Derive a [program address][pda] from the given seeds, optional bump and
+/// program id.
+///
+/// [pda]: https://solana.com/docs/core/pda
+///
+/// In general, the derivation uses an optional bump (byte) value to ensure a
+/// valid PDA (off-curve) is generated. Even when a program stores a bump to
+/// derive a program address, it is necessary to use the
+/// `Address::create_program_address` to validate the derivation. In
+/// most cases, the program has the correct seeds for the derivation, so it would
+/// be sufficient to just perform the derivation and compare it against the
+/// expected resulting address.
+///
+/// This function avoids the cost of the `create_program_address` syscall
+/// (`1500` compute units) by directly computing the derived address
+/// calculating the hash of the seeds, bump and program id using the
+/// `sol_sha256` syscall.
+///
+/// # Important
+///
+/// This function differs from `Address::create_program_address` in that
+/// it does not perform a validation to ensure that the derived address is a valid
+/// (off-curve) program derived address. It is intended for use in cases where the
+/// seeds, bump, and program id are known to be valid, and the caller wants to derive
+/// the address without incurring the cost of the `create_program_address` syscall.
+pub fn derive_address<const N: usize>(
+    seeds: &[&[u8]; N],
+    bump: Option<u8>,
+    program_id: &Address,
+) -> Address {
+    #[cfg(target_os = "solana")]
+    {
+        const {
+            assert!(N < MAX_SEEDS, "number of seeds must be less than MAX_SEEDS");
+        }
+
+        const UNINIT: MaybeUninit<&[u8]> = MaybeUninit::<&[u8]>::uninit();
+        let mut data = [UNINIT; MAX_SEEDS + 2];
+        let mut i = 0;
+
+        while i < N {
+            // SAFETY: `data` is guaranteed to have enough space for `N` seeds,
+            // so `i` will always be within bounds.
+            unsafe {
+                data.get_unchecked_mut(i).write(seeds.get_unchecked(i));
+            }
+            i += 1;
+        }
+
+        // TODO: replace this with `as_slice` when the MSRV is upgraded
+        // to `1.84.0+`.
+        let bump_seed = [bump.unwrap_or_default()];
```

Comment:
It looks like `as_slice()` is stable in non-const contexts as of 1.75: https://doc.rust-lang.org/std/option/enum.Option.html#method.as_slice, so from what I see, this should be safe to change, no?

Reply:
Good catch! We only need `1.84+` for the `const` implementation.